paddle/phi/core/generator.cc

```diff
@@ -157,134 +158,125 @@ const std::shared_ptr<Generator>& GetRandomSeedGenerator(
+Generator::Generator(uint64_t seed, int64_t device_id) {
+  current_index = states_.size();
+  // device id first, then seed
+  states_.emplace_back(device_id, seed);
+  print_state_info();
 }
 
-phi::Generator::GeneratorState Generator::GetState() {
-  std::lock_guard<std::mutex> lock(this->mu_);
-  state_.cpu_engine = *engine_;
-  VLOG(4) << "Get Random state: "
-          << "device id: " << (uint64_t)(this->state_.device)
-          << ", current_seed: " << this->state_.current_seed
-          << ", thread_offset: " << this->state_.thread_offset
-          << ", cpu engine: " << *(this->engine_);
-  return this->state_;
-}
+phi::Generator::GeneratorState Generator::GetState() { return state().clone(); }
 
 void Generator::SetState(const phi::Generator::GeneratorState& state) {
-  std::lock_guard<std::mutex> lock(this->mu_);
-  this->state_ = state;
-  this->engine_ = std::make_shared<std::mt19937_64>(state.cpu_engine);
-  VLOG(4) << "Set Random state: "
-          << "device id: " << (uint64_t)(this->state_.device)
-          << ", current_seed: " << this->state_.current_seed
-          << ", thread_offset: " << this->state_.thread_offset
-          << ", cpu engine: " << *(this->engine_);
+  std::lock_guard<std::mutex> lock(mu_);
+  if (current_index < states_.size())
+    states_[current_index] = state.clone();
+  else
+    PADDLE_THROW(phi::errors::NotFound("Generator index is not found"));
+  print_state_info();
+}
+
+uint64_t Generator::GetStateIndex() { return current_index; }
+
+void Generator::SetStateIndex(uint64_t StateIndex) {
+  std::lock_guard<std::mutex> lock(mu_);
+  if (current_index < states_.size())
+    current_index = StateIndex;
+  else
+    PADDLE_THROW(phi::errors::NotFound("Generator index is not found"));
+}
+
+uint64_t Generator::RegisterStateIndex(const GeneratorState& state) {
+  std::lock_guard<std::mutex> lock(mu_);
+  auto new_index = states_.size();
+  states_.push_back(state);
+  current_index = new_index;
+  return new_index;
+}
+
+inline Generator::GeneratorState& Generator::state() {
+  if (current_index < states_.size())
+    return states_[current_index];
+  else
+    PADDLE_THROW(phi::errors::NotFound("Generator index is not found"));
+}
+
+inline std::shared_ptr<std::mt19937_64> Generator::cpu_engine() {
+  return state().cpu_engine;
 }
 
 uint64_t Generator::GetCurrentSeed() {
-  std::lock_guard<std::mutex> lock(this->mu_);
-  return this->state_.current_seed;
+  std::lock_guard<std::mutex> lock(mu_);
+  return state().seed;
 }
 
 uint64_t Generator::Seed() {
-  std::lock_guard<std::mutex> lock(this->mu_);
-  uint64_t seed = 0;
-  std::random_device de;
-  seed = ((((uint64_t)de()) << 32) + de()) & 0x1FFFFFFFFFFFFF;
-  this->state_.current_seed = seed;
-  std::seed_seq seq({seed});
-  this->engine_->seed(seq);
-
-  return this->state_.current_seed;
+  std::lock_guard<std::mutex> lock(mu_);
+  uint64_t seed = GetRandomSeed();
+  state().reset(seed);
```

Comment:
The code logic here is different from the original ones. The original one does not reset `this.state_.offset` but the new one reset `state().offset`.

Reply:
The reset of offset in the new implementation is an intentional change. When seed() is invoked, it implies the generation of a new random value for the seed and setting the current seed to this random value. Consequently, every time a seed is set, it is logical and necessary to reset the value of offset as well. This approach aligns with [`SetCurrentSeed`](https://github.com/PaddlePaddle/Paddle/blob/v2.6.1/paddle/phi/core/generator.cc#L238) in our repository and the implementation in PyTorch [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cuda/CUDAGeneratorImpl.cpp#L163), where the offset is reset whenever a new seed is set.